salt/orch/update.sls

```diff
@@ -238,10 +228,25 @@ etcd-setup:
     - kwarg:
         destructive: True
     - require:
-      - {{ master_id }}-post-start-services
+      - {{ master_id }}-start-services
 
 {% endfor %}
 
+# Perform migrations after all masters have been updated
+all-masters-post-start-services:
+  salt.state:
+    - tgt: '{{ is_master_tgt }}'
+    - tgt_type: compound
+    - batch: 3
+    - sls:
+      - cni.update-post-start-services
+      - kubelet.update-post-start-services
+    - require:
+      - etcd-setup
+{%- for master_id in masters.keys() %}
+      - {{ master_id }}-reboot-needed-grain
+{%- endfor %}
```

Comment:
What is the reason for doing the `{cni|kubelet}-post-start-services` in the master in parallel instead of one-by-one (as we have been doing so far)?

Reply:
In this case, the upgrade from 2.0 to 3.0 includes an [auth change](https://github.com/kubic-project/salt/pull/424/files#diff-7504d75a43d2c611b6f390b9c99a27d6). This means that if we have M1, M2 and M3. We update M1 first, when it reaches the `post-start-services`, it will try to run `kubectl delete node old_node_name`, but that request can land in M2 or M3, that doesn't yet understand the new auth method (didn't get updated yet, we are updating M1), so the removal of the old nodename does not work.

This makes sure we perform this task after all master machines have been upgraded. The workers keep the same logic one by one, as at that point in time all masters have been updated and they all understand the authentication method.

Reply:
I see, thanks for the clarification...